### PR TITLE
Support INTO clauses in AQL hashed COLLECT variant

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Backport of https://github.com/arangodb/arangodb/pull/18252

* Make the hashed variant of AQL COLLECT support INTO clauses too. Previously only the sorted variant of AQL COLLECT supported INTO clauses.

Helps https://arangodb.atlassian.net/browse/PRESUPP-546

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [x] Backport for 3.10: this PR
  - [ ] Backport for 3.9: -
  - [ ] Backport for 3.8: -

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

**Review by mpoeter — APPROVED** [review]

**Review by mchacki — CHANGES_REQUESTED** [review]
Production code looks good to me.

But can we have tests for this feature?

All the tests I have seen is that the Optimizer picks the change up and rejects what is incorrect (good)
But there is no "happy path" test, that actually asserts that the behavior of this node is as expected.

File: tests/js/server/aql/aql-optimizer-collect-into.js [hunk 1]
```diff
@@ -49,185 +43,241 @@ function optimizerCollectExpressionTestSuite () {
...
+          assertTrue(row.docs[i].value >= 0 && row.docs[i].value < 1);
+        }
+
+        let nodes = AQL_EXPLAIN(query).plan.nodes.filter((n) => n.type === 'CollectNode');
+        assertEqual(1, nodes.length);
+        assertEqual(method, nodes[0].collectOptions.method);
+      });
+    },
+    
+    testCollectMethodsIntoAndAgreegate : function () {
```
Comment (mchacki) [inline tests/js/server/aql/aql-optimizer-collect-into.js:122]:
```suggestion
    testCollectMethodsIntoAndAggregate : function () {
```

File: tests/js/server/aql/aql-optimizer-collect-methods.js [hunk 2]
```diff
@@ -45,46 +41,77 @@ function optimizerCollectMethodsTestSuite () {
       db._drop("UnitTestsCollection");
       c = db._create("UnitTestsCollection", { numberOfShards: 3 });
 
-      for (var i = 0; i < 1500; ++i) {
-        c.save({ group: "test" + (i % 10), value: i, haxe: "test" + i });
+      let docs = [];
+      for (let i = 0; i < 1500; ++i) {
+        docs.push({ group: "test" + (i % 10), value: i, haxe: "test" + i });
```
Comment (mchacki) [inline tests/js/server/aql/aql-optimizer-collect-methods.js:46]: Like it :+1:

**Review by mchacki — APPROVED** [review]
LGTM.
Minor typo in test name, happy to keep it though...

**Comment by jsteemann** [comment]
Marked as "DO_NOT_MERGE" because with the PR, execution plans may change in the following way:
* we may have more (extra) execution plans compared to previous 3.10 versions
* the "hash" variant of COLLECT may use more memory than the "sorted" variant
So it is unclear if we want to land this PR in the middle of a stable release. Potentially it is better to introduce the change only in the next version (3.11). 
@Zyqsempai

**Review by Zyqsempai — CHANGES_REQUESTED** [review]
@jsteemann Let's not backport it to 3.10, considering your concerns and the fact that it could wait till 3.11 release.

**Comment by jsteemann** [comment]
Closed because feature shouldn't be backported.

**Pull request closed without merging**